### PR TITLE
Improve component testing interface

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?

This PR proposes a new, less verbose way to write component unit test. 

The idea is: 
instead of `fxutil.Test` fulfilling a closure containing an arbitrary list of dependencies, we instead return a generic container with it's dependencies already fulfilled. 

This assumes that a test file will define a struct of dependencies, and the test itself will specialize the generic `Test` function such that it returns the fulfilled dependencies. 

So your test that used to look like:
```go 
func test(t *testing.T) {
	fxutil.Test(t, fx.Options(...), func(a CompA, b CompB, c CompC) {
		// test a
		// test b 
		// test c
	}
}
```
will now look like this

```go
type testDeps struct {
    A CompA
    B CompB
    C CompC
}

func test(t *testing.T) {
	deps := fxutil.Test[testDeps](t, fx.Options(...))
	// test deps.A
	// test deps.B
	// test deps.C
}
```

This also works for individual components:

```go

func test(t *testing.T) {
	compA := fxutil.Test[CompA](t, fx.Options(...))
	// test compA
}
```

This has several advantages: 

1. Go functions do not support optional arguments, so writing a test with a different dependency list requires defining a whole new function, or explicitly ignoring the argument.
2. Removes an unnecessary level of nesting which makes tests harder to reason about. 
3. Makes test utility functions more reusable. 

[You can see a complete example of the old approach here](https://github.com/DataDog/datadog-agent/blob/main/comp/dogstatsd/server/server_test.go#L98) which should hopefully illustrate the challenges. 

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by hush-hush — APPROVED** [review]
This looks great ! I love the removal of  the callback, it simplifies a lot the tests.

File: comp/process/runner/runner_test.go [hunk 1]
```diff
@@ -25,8 +25,13 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
 )
 
+type testDeps struct {
+	fx.In
+	Runner Component
+}
```
Comment (ogaca-dd) [inline comp/process/runner/runner_test.go:31]: I think using only `Component` should be enough

**Review by ogaca-dd — APPROVED** [review]
LGTM

File: comp/process/runner/runner_test.go (hunk 1 above)
Comment (gh123man) [reply comp/process/runner/runner_test.go:31]: Oops! missed one :) thanks

**Review by just-chillin — APPROVED** [review]
Looks great! Thanks for doing this! Love seeing more uses of generics in go

**Review by scottopell — APPROVED** [review]

**Review by truthbk — APPROVED** [review]

**Pull request merged**